fairroot/base/steer/FairRootManager.cxx

```diff
@@ -120,6 +114,30 @@ FairRootManager::~FairRootManager()
     LOG(debug) << "FairRootManager::~FairRootManager done #" << fId << " at " << this;
 }
 
+void FairRootManager::SetSource(FairSource* source)
+{
+    fSource = source;
+    fSourceOwned.reset(source);
+}
+
+void FairRootManager::SetSourceNonOwning(FairSource* source)
+{
+    fSource = source;
+    fSourceOwned.reset();
+}
+
+void FairRootManager::SetSink(FairSink* sink)
+{
+    fSink = sink;
+    fSinkOwned.reset(sink);
+}
+
+void FairRootManager::SetSinkNonOwning(FairSink* sink)
+{
+    fSink = sink;
+    fSinkOwned.reset();
+}
+
```

Comment:
If the input parameter is a raw pointer, according to the common convention, it already means a non-owning. So I would say the `NonOwning` in the function name is unnecessary. 

 Second, since fSink is owned by the FairRootManager, its setter should be owning:
```cpp
std::unique_ptr<FairSink> fSink;
void FairRootManager::SetSink(std::unique_ptr<FairSink> sink)
{
     fSink = std::move(sink);
}
[[deprecated]]void FairRootManager::SetSink(FairSink* sink)
{
     fSink.reset(sink);
}
```

I also don't quite understand why we have two member variables for the sink in this case.

Reply:
First things first:

The idea is to let FairRun own the sink and source. FairRun::Set{Sink,Source} should be the APIs for users. FairRootManager only needs a reference to the sink and source. So `FairRootManager::Set{Sink,Source}*` *should* be internal and only take non-owning pointers.

And yes, I should add that to the commit body and PR message above. Will do so next week after #1432 gets merged.

That said, we have a problem:
The current FairRootManager::Set{Sink,Source}` member functions are public and not marked as internal. So people are allowed to use them (and I think, they're used). Those APIs take a half-owning pointer currently (which you correctly diagnosed in your issue!). This is quite broken, really, and should be deprecated.
So:
* We need to keep the old (broken) API for a bit of time for the current users.
* Introduce a new non-owning, internal API for FairRun.

We need a name for the new API. I made up this Set{Sink,Source}NonOwning() name. I am fully open for better names. Maybe something with "internal" in it? I don't know.

Reply:
Sorry for the late reply. Maybe we should just deprecate the setters of FairRootManager, without adding any new setters for the sink and source?

> So FairRootManager::Set{Sink,Source}* should be internal and only take non-owning pointers.

There shouldn't be a setter even for the internal usage. Sink and source should be set directly from users, not from anything else. If there is a internal way to set the sink or source, users basically would have no idea about it and this could cause a lot of problems in the future.

Reply:
FairRootManager is a mess. We're trying to reduce the messiness somewhat. For this PR I am trying to fix the double-delete and not introduce too much messiness while still keeping (and deprecating) public APIs.

For this discussion, let's please assume that FairRootManager is itself an internal helper class (or at least it should be) that needs to have a reference to a sink and/or source to be able to fulfil its job. To set this reference … well … I think, we should use a setter?

As FairRootManager isn't yet an internal only helper class, we need to make that setter "internal". Either by adding docs about it being internal, or by making it `private` and using `friend` to allow FairRun to call it. I somewhat prefer comments.

Reply:
> let's please assume that FairRootManager is itself an internal helper class (or at least it should be) that needs to have a reference to a sink and/or source to be able to fulfil its job.

There could be a lot of problems if we have a source/sink member variables (either by pointer or reference) and the "internal setters" in FairRootManager. The biggest one is about the lifetime of source/sink. Since they are not owned by FairRootManager, they could be deleted by FairRun and then accessed by the FairRootManager (, which causes a seg fault). A good design should disallow this to happen.

So I would suggest to completely remove source/sink related (public & private) member variables and functions. If FairRootManager wants to use it, simply get it from FairRun, which owns the source/sink resource:
```cpp
void FairRootManager::MyFunction()
{
     auto* source = FairRun::Instance()->GetSource();
     UseThisSource(source);
}
```
With this, the lifetime of source/sink resources can always be guaranteed.

Reply:
`FairRun::Instance()` references global state. We want to reduce global state in the long run, really.

That said, let's look at lifetime.

First: Thanks for poking on this one! I fixed an issue now.

So the main risk is that FairRootManager (FRM) has a dangling reference, because FairRun deleted the sink/source. All the member functions that get a new sink/source properly reset the FRM thing before `reset()`in the unique_ptr (and thus deleting the old sink/source. So those should be safe. I missed the destructor! So now the destructor also resets the FRM sink/source, before the member variable destructors of `FairRun::{fSink,fSource}` delete the sink/source. So this should be safe as well now.

Reply:
>All the member functions that get a new sink/source properly reset the FRM thing before reset()in the unique_ptr (and thus deleting the old sink/source. So those should be safe. I missed the destructor! So now the destructor also resets the FRM sink/source, before the member variable destructors of FairRun::{fSink,fSource} delete the sink/source. So this should be safe as well now.

Hi, 

I would say using this trick just makes sure the algorithm can run successfully and it doesn't guarantee the safety of the code. For example, what if FairRootManager needs to write data to the fSink and FairRun is deleted before this action takes place? We could have the program running successfully but it fails to produce expected result. And a successfully running program is even harder to debug than a program that crashes. For example, using `auto* run = new FairRun()` (forgetting to delete it in the end) and `auto run = std::make_unique<FairRun>()` can get totally different results.

> We want to reduce global state in the long run, really.

Generally, I am not fan of neither `reset` or singleton. We don't have control of when the reset functions are called. And bad things could happen when they are called either too early or too late or not being called at all. In case of singleton, we can't have control over when constructors or destructors are called.

Like I have mentioned above, there is a better way to manage this: **chain of responsibility**. If sink/source files are owned by FairRun, everything related to them should be handled via their owner, FairRun. If FairRootManager needs to write the data to sink file, all it needs to do is to own FairRun and pass the FairRun pointer to the callee. In this case, FairRootManager doesn't deal with sink and source file directly. And therefore, it doesn't need to care about the lifetime of sink or source file anymore. It only needs to care about whether the owning FairRun is still alive or not.

Reply:
TBH, I think, most of your points are quite academical now.

> For example, what if FairRootManager needs to write data to the fSink and FairRun is deleted before this action takes place?

To let this happen, one has to create a quite complex setup:

* FairRootManager is a `thread_local`. So each thread has one.
* main thread (prepare):
  * Setup FairRootManager& `FRM`
  * Setup FairRun* `run`.
  * `run->SetSink(std::move(somesink));`
  * calls `FRM.SetSinkNonOwning`.
  * We're ready.
* main thread: Fire up a new thread "deleter", hand over `run`
* "deleter" thread needs time to start up / sleeps a bit.
* main thread: start to call `FRM.SomeSinkWriter`. It loads `fSink` into a register, and sleeps
* "deleter" thread now finally `delete run;`. sink goes away. (and it calls `FRM.SetSinkNonOwning(nullptr)` (on the "right" FRM, because FairRun has a reference to the FRM in the other thread), but as the reference to the sink is already in a register, that's not important.
* main thread awakes and uses the sink in the register to act on the destructed sink.
* Booom.

Yes, this is possible.

I *think* that FairRoot has way more problematic thread issues than this one. In other words: If a user uses threads manually (like in this case), then they should be very careful on their own.

Did I miss anything?

Reply:
I think, most of this discussion is now mostly out dated with the current maybe_owning_ptr implementation. Or it's highly academic concerning the current real world use cases.

Yes, the implementation is problematic. But it's less problematic then the current state. And it `[[deprecated]]`s some current problematic cases anyway.